File: assets/csv_to_map.py
import csv

if __name__ == '__main__':
    tiles = []

    with open('big.csv') as csvfile:
        csv_reader = csv.reader(csvfile)
        height = 0
        for row in csv_reader:
            height += 1
            width = 0
            for tile in row:
                width += 1
                tiles.append(int(tile))

    map_size = width * height
    if map_size > 0xffff:
        exit('map to large')
    ms_b1, ms_b2 = (map_size & 0xffff).to_bytes(2, 'big')

    print(hex(ms_b2), hex(ms_b1), hex(width), hex(height))
    raw_tilemap = [ms_b2, ms_b1, width, 00, height, 00] + tiles
    raw_tilemap + [0] * (256 - len(raw_tilemap) % 256)

    tilemap = bytearray(raw_tilemap)
    with open('big.bin', 'wb') as out_file:
        out_file.write(tilemap);

File: assets/csv_to_map.rb
require 'csv'

# map format
# header bytes
# 0-1: map size
# 2: map height
# 3: map width
# 4: map[0]

def hex(num, rjust_len = 2)
    (num || 0).to_s(16).rjust(rjust_len, '0').upcase
end

filename = ARGV[0]
height = 0
width = nil
tiles = []

raise "File #{filename} not found" unless File.file?(filename)

CSV.foreach(filename) do |row|
    height += 1
    width ||= row.length
    row.each do |cell|
        tile = cell.to_i
        raise "Tile value too big" if tile > 0xff
        tiles << tile
    end
end

raise "Width too big" if width > 0xff
raise "Height too big" if height > 0xff

size = height * width
raise "Map too big" if size > 0xffff

size_msb = size >> 8
size_lsb = size & 0xff

tiles = [size_lsb, size_msb, height, width] + tiles

File.open("#{filename}.bin", "w+b") do |file|
    file.write([tiles.map { |i| hex(i) }.join].pack('H*'))
end
